### PR TITLE
Remove intermediate InsertNodeAtPath in ipfs add

### PR DESCRIPTION
**Pull request by rht**
This is an experiment.
The time it takes for an `ipfs add` is halved when the intermediate root(s) are not created.

### CODE REVIEW

**Comment by rht** [comment]
no-sync with intermediate root
outdata

no-sync with no intermediate root
noinsertnode

**Comment by whyrusleeping** [comment]
makes a lot of sense to me. That way when adding directories, we dont add the files under them multiple times. Good catch.

**Comment by whyrusleeping** [comment]
@rht what script are you using to generate those graphs? I'd like to have it too :)

**Comment by rht** [comment]
@whyrusleeping The slow part is in looking up the subdir nodes `root.GetLinkedNode`. Duplicated `dagservice.Add` usually ends up as a check whether the block already exists (this can be optimised still). Updated the script to fix darcs add. As you can see, `ipfs add` is faster than `darcs add` without `InsertNodeAtPath`.

https://ipfs.io/ipfs/QmUjUKKtAEKecKmvS8TznxRBPj3qfTERea5RnUY7HFUHPc (requires `python-numpy` and `python-matplotlib`).

**Comment by rht** [comment]
Removing intermediate root(s) also means removing intermediate hashing.
The hash of the said root(s) are never broadcasted anyway. Only the subnode hashes are broadcasted in `outputDagNode`.

**Comment by rht** [comment]
Next bottleneck that can be optimized is https://github.com/ipfs/go-ipfs/blob/master/importer/balanced/builder.go#L12 (potentially accounts for 20-25% of the time taken).

The incoming chunks are channeled, but is then processed serially in `BalancedLayout`.

**Comment by rht** [comment]
If the chunks are not channeled (no-sync, no-insertnode, chunk-not-channeled, 1KB each), the perf is approaching git
outdata

**Comment by whyrusleeping** [comment]
@rht do you have the 'chunk-not-channeled' code around somewhere?

**Comment by rht** [comment]
In here https://github.com/rht/go-ipfs/tree/no-chunk-channel (also I found an old benchmark on 'channel as iterator' https://groups.google.com/forum/#!msg/golang-nuts/pxfhKGqHNv0/If4Gz09r_2gJ if this can be run in 1.5).

**Comment by rht** [comment]
`outputDagnode` takes a while, as can be seen when adding go-ipfs' Godeps, the time it takes for the entire `AddFIle` in core/commands/add.go: no-insertnodeatpath (440ms) -> no-chunk-channel (254ms) -> no-outputDagnode (188ms).

**Comment by rht** [comment]
There are 2 kinds of benchmark I used:
1. how `ipfs add` scale when adding lots of small files + plot
2. `ipfs add -q -r Godeps >actual` while measuring the total time a function spends (adding `now := time.Now()` before the statement, and `fmt.Println(time.Since(now).Nanoseconds()` after, then summing all the times with [1]).

To reproduce the number in the previous comment, I use benchmark 2.

[1]
```python
    print float(sum(int(i) for i in open('actual').read().split('\n') if i and (not i.startswith('Qm'))))/1000000, 'ms'
```

**Comment by rht** [comment]
This is the fastest so far (no-insertnodeatpath + no-chunk-channel + no-outputDagnode)
outdata

**Comment by whyrusleeping** [comment]
@rht does buffering the channel in question have a similar effect on the runtime?

**Comment by rht** [comment]
Will try, but just discovered this is faster yet (~O(rsync), no-postrun-outchan-loop, no-prerun-size)
outdata

**Comment by whyrusleeping** [comment]
whoa, thats some crazy jitter there past 3500 files.

**Comment by whyrusleeping** [comment]
and just for the hell of it, you should throw in the time for a `dd` of equivalent size to all files being added, lol

**Comment by rht** [comment]
(I did it with the `random-files` line)

The last one is a hack, I removed the `go func() {addAllAndPin...}()` and it is back at
outdata

Slightly faster than the previous case (likely because the file tree isn't initially traversed), but hasn't passed the git barrier yet (and that it is not impossible).

**Comment by rht** [comment]
Buffered channels (no-sync + no-insertnodeatpath + buffered-channel, 1KB each)
outdata

not much if compared to the serialized version https://github.com/ipfs/go-ipfs/pull/1964#issuecomment-156788359.

**Comment by whyrusleeping** [comment]
yeah, i'm liking the serialized version. didnt realize the perf hit was that high on channels o.o

**Comment by rht** [comment]
Here is the 'channel as iterator' run as of go1.5.1 (from https://groups.google.com/forum/#!msg/golang-nuts/pxfhKGqHNv0/If4Gz09r_2gJ):

- simple loop 0.39s
- counter object 2.33s
- goroutine 33.27s
- goroutine buffered 16.62

**Comment by whyrusleeping** [comment]
yeah, i was reading that. it makes sense i guess

**Comment by rht** [comment]
Reran the test. The `random-files` err from previous test disappeared.
Current err is travis err with fuse mount, wasn't present previously.

**Comment by jbenet** [comment]
> didnt realize the perf hit was that high on channels o.o

every gopher eventually

**Comment by jbenet** [comment]
@rht this is very good. a bit hard to follow the changes without in depth look, for my speedier CR would you mind describing what the old algo was doing, and what the new one does?

It may be useful, btw, to create an implementation of Node for dag editing that does not serialize or hash anything during edits to the graph and lazily does so when the node's serialized fmt (or hash) is requested. I.e. keeps pointers to objects until the last moment possible. -- though no sense spending time getting that to work before the impending ipld landing.

**Comment by rht** [comment]
I have decoupled no-sync from the PR.

With the old algo, the memory dag service creates and accumulates all the intermediate incomplete roots (where in #1965, the old intermediate root is removed, that only one root node exists after each operation). In this PR, no intermediate roots are created/accumulated, only the subroots that build up to the intended final root.

**Comment by whyrusleeping** [comment]
@rht in the future, could you push your branches to this repo? makes it a little easier on my end for testing things out

**Comment by whyrusleeping** [comment]
@rht pulled this down for some local testing (after merging #1965 ). Running tests on btrfs, 500MB/s R/W ssd, and 32GB ram.

The test i ran was adding a directory containing 10,000 1MB files. Before each test i deleted my ipfs repo, ran ipfs init, and started the daemon.

On dev0.4.0 (latest, with 1965), it took roughly 7 minutes (+/- 10s) to complete.
On this branch, it was not noticeably different (within one standard deviation).
With no-sync merged in and enabled (into dev0.4.0)  the test ran in ~55 seconds (+/- 2s)
With no-sync merged in and enabled on this branch, the test ran in ~49 seconds (+/- 2s)

I thought you mentioned that no-sync no longer had a significant effect on the runtime? did i miss something?

**Comment by whyrusleeping** [comment]
I noticed that #1973 was some of the other changes separated out, running just that branch also produced similar results to plain dev0.4.0 (roughly 7min, best time 6:48, worst 7:02, out of four runs)

**Comment by rht** [comment]
Ok to be specific, the test I ran was sync + (no-insertnodeatpath + no-chunk-channel).
Currently rechecking the result again.

**Comment by whyrusleeping** [comment]
merging those two together didnt appear to have an effect on the runtime (granted, i didnt let them finish and just used the estimate for a quick check that time).

My git history to show what code i'm running:
```
6c99c2e Merge branch 'rht-no-chunk-channel' into rht-no-insertnodeatpath
2916c67 Merge branch 'no-chunk-channel' of https://github.com/rht/go-ipfs into rht-no-chunk-channel
40d6ada Merge branch 'no-insertnodeatpath' of https://github.com/rht/go-ipfs into rht-no-insertnodeatpath
5f8cfc9 Merge pull request #1965 from ipfs/fix/add-mem-4
3f21b8a Simplify BalancedLayout test
0234b83 Remove chunk channels
6e12c53 Remove intermediate InsertNodeAtPath in ipfs add
02e14c4 comment multiple dagstore error checking
09d34db Merge pull request #1931 from eminence/addnopin_v2
e002194 rework editor creation and finalization
605b24f improves memory usage of add
b4c4591 Add a --pin option to `ipfs add` (allowing --pin=false)
1bbc472 Move parts of `ipfs add` into core/coreunix
```

**Comment by rht** [comment]
ok, I reproduced the "fast sync" case once again.
I combined all the three branches (add-no-sync, no-insertnodeatpath, no-chunk-channel), explicitly set the NoSync flag to false, then it resulted in a fast add again.

**Comment by rht** [comment]
The flag must been set to true somehow regardless of the config?

**Comment by rht** [comment]
Another check: I hardcoded the syncfs flag to be true, still get the "fast sync" result. The problem is in the go-datastore then.

**Comment by rht** [comment]
Found it: in `flatfs.New`, the sync argument doesn't get passed into the Datastore object creation.

**Comment by whyrusleeping** [comment]
@rht wow, good catch. thanks!

**Comment by jbenet** [comment]
LGTM. ok so do we still want to merge this @whyrusleeping ? 

(what if the "edited dag" is too big to hold in memory? not sure what happens now? fails? but in future should store on disk and gc memory re-reading as necessary? (should be possible to edit massive dags with very little memory, but enough disk))

**Comment by rht** [comment]
(rebased)

The last / only InsertNodeAtPath in this PR doesn't consume much memory, since what it does is just patching the final hash to the "global" root of an IpfsNode (as in, the root dir one would see of /ipfs if the repo is fuse-mounted).

**Comment by whyrusleeping** [comment]
closing, we use the mfs code for adds now (if changes from here are still relevant to that process, please file a new PR on top of master with them)
